src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,77 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(DataFrames.nonunique(df[:,indexcols]))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
+    colind = index(df)[indexcols]
+
+    dummydf = similar(select(df, indexcols, copycols=false),0)
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if sum(ismissing.(df[col])) != 0
+            tempcol = vcat(levels(df[col]), missing)
+        else
+            tempcol = levels(df[col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(flatten(product(values(uniqueVals)...)))
+    dfsize = div(length(collected),length(indexcols))
+
+    for i in 1:length(indexcols)
+        _expandhelper(dummydf[!,i], collected, length(indexcols), i)
+    end
+
+    return dummydf
+end
+
+function complete(df::AbstractDataFrame, indexcols::Array{Symbol,1}; fill=missing::Any, replaceallmissing=false)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(DataFrames.nonunique(df[:,indexcols]))>0 && throw(ArgumentError("duplicate rows in input"))
+    colind = index(df)[indexcols]
+
+    # Expand the input df and left join
+    expanded = expand(df, indexcols)
+    expanded = join(expanded, df; on = indexcols, kind = :left, indicator = :source)
+
+    # Replace missing values with the fill
+    if !ismissing(fill)
+        for n in names(expanded)
+            if (n ∉ indexcols) && (n != :source)
+                if replaceallmissing == true
+                    expanded[n] = coalesce.(expanded[n],fill)
+                else
+                    #expanded[n] .= ifelse.(expanded.source .== "left_only", fill, expanded[n])
+                    let col = expanded[n]
```

Comment:
inside a function `let` does not help - you need a helper function to benefit from barrier function approach as above. `let` is a trick for global scope.

Reply:
also `expanded[!, n]`

Reply:
Tried my hand at barrier-izing it after reviewing the julia performance tips docs you linked me before. Let me know if I did it right and can resolve this!

Reply:
OK - I will go through the whole PR and let you know if I notice something.